src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -426,6 +461,60 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
     }
   }
 
+  function findClosestErrorBoundary(fiber : Fiber): ?Fiber {
+    let maybeErrorBoundary = fiber.return;
+    while (maybeErrorBoundary) {
+      if (maybeErrorBoundary.tag === ClassComponent) {
+        const instance = maybeErrorBoundary.stateNode;
+        if (typeof instance.unstable_handleError === 'function') {
+          return maybeErrorBoundary;
+        }
+      }
+      maybeErrorBoundary = maybeErrorBoundary.return;
+    }
+    return null;
+  }
+
+  function handleError(failedUnitOfWork : Fiber, error : any) {
+    const errorBoundary = findClosestErrorBoundary(failedUnitOfWork);
+    if (errorBoundary) {
+      handleErrorInBoundary(errorBoundary, error);
+      return;
+    }
+    // TODO: Do we need to reset nextUnitOfWork here?
+    throw error;
+  }
+
+  function handleErrorInBoundary(errorBoundary : Fiber, error : any) {
+    // The work below failed so we need to clear it out and try to render the error state.
+    // TODO: Do we need to clear all of these fields? How do we teardown an existing tree?
```

Comment:
If we naively just clear everything here, we'll lose the ability to unmount them. I don't think we need to clear anything here.

We normally we reset everything to whatever the "current" tree represents on the way down the tree. The simplest way might be to restart from the root.

In theory the error boundary will already be a clone, it might be fine to just restart from it. However, `unstable_handleError` may be calling `setState` on a component above (or even a sibling) and that will start screwing everything up. If you restart from the root, that case will be covered.